### PR TITLE
[Enhance] concat micro batch inputs before emb's and lm_head's forward

### PR DESCRIPTION
**Pull request by HIT-cwh**
Concat the micro-batch inputs to ensure the all-gather operation is executed only once.

### CODE REVIEW

File: xtuner/v1/data_proto/sequence_context.py [hunk 1]
```diff
@@ -199,7 +199,7 @@ def split(self, sequence_parallel_mesh: DeviceMesh | None = None) -> Self:
             return self
 
     @classmethod
-    def pack(cls, sequence_context_list: list["SequenceContext"]):
+    def cat(cls, sequence_context_list: list["SequenceContext"]):
```
Comment (HAOCHENYE) [inline xtuner/v1/data_proto/sequence_context.py]: Return annotation

**Comment by HAOCHENYE** [comment]
##  Overview (Reviewed by Claude Code)

  This PR refactors the MoE (Mixture of Experts) model's micro-batch forward pass to improve efficiency by:
  1. Renaming pack() → cat() for SequenceContext (more accurate naming)
  2. Adding a cat() method to BaseLossKwargs and BaseLossContext classes
  3. Major optimization: Consolidating tensor operations in _micro_batch_forward() to reduce redundant computations

 ## Analysis

  ✅ Positive Changes

  1. Method Naming Improvement (sequence_context.py)
    - Renaming pack → cat is more semantically accurate since the method concatenates tensors using torch.cat
  2. New cat() Methods (base_loss_ctx.py)
    - Well-structured implementation that mirrors the existing chunk() method
    - Correctly uses dim=1 to reverse the chunk operation
    - Proper assertions for input validation
  3. Performance Optimization (moe.py)
    - The old code computed embeddings per micro-batch in a loop, then concatenated later
    - New code concatenates input tensors first, then does a single embedding lookup
    - This reduces kernel launch overhead and improves GPU utilization

##  ⚠️  Potential Issues & Suggestions

  1. Variable Shadowing (moe.py:419)
  loss, (logits, extra_info) = self.lm_head(cat_hidden_states, cat_loss_ctx)
  # ...
  loss: torch.Tensor  # This type annotation shadows the variable
  output["loss"] = loss.sum()
    - The loss: torch.Tensor annotation is redundant and potentially confusing
    - Suggestion: Remove the type annotation line
  2. Unused Variable (moe.py:427-428)
  final_logits = logits
    - final_logits is assigned but the logic for handling it was removed. Need to verify it's still used downstream.
  3. Logic Change Risk (moe.py:406-408)
  cat_hidden_states = torch.cat(hidden_states_list, dim=1)
  cat_hidden_states = self.norm(cat_hidden_states)
    - This concatenates hidden_states_list again after it was already modified in the MoE branch
    - The old code processed normalization per micro-batch; new code does it on concatenated tensor
    - Question: Are these mathematically equivalent? LayerNorm across longer sequences vs. per micro-batch may differ slightly
  4. Missing Assertion (moe.py:350-351)
  if seq_ctx_list[0].input_ids is None:
    - No check that seq_ctx_list is non-empty before accessing [0]
    - Suggestion: Add assertion or guard
  5. Comment Language Consistency (base_loss_ctx.py)
    - Comments use Chinese (收集所有 tensor 字段名...)
    - Consider using English for consistency with the rest of the codebase

##   🔍 Questions for Author

  1. Has this been benchmarked to confirm the performance improvement?
  2. Are there unit tests covering the new cat() methods?
  3. Has the numerical equivalence been verified (especially for the LayerNorm change)?

  Summary

  | Aspect           | Assessment                                                                    |
  |------------------|-------------------------------------------------------------------------------|
  | Code Correctness | ⚠️  Minor concerns (variable shadowing, potential semantic difference in norm) |
  | Code Style       | ✅ Good overall, minor comment language inconsistency                         |
  | Performance      | ✅ Should improve through batched operations                                  |
  | Test Coverage    | ❓ Unknown - tests not visible in diff                                        |
  | Security         | ✅ No concerns                                                                |

  Recommendation: Approve with minor revisions - address the variable shadowing and verify numerical equivalence of the LayerNorm change.

File: xtuner/v1/model/moe/moe.py [hunk 2]
```diff
@@ -353,28 +348,22 @@ def _micro_batch_forward(
...
+                    cat_seq_ctx = SequenceContext.cat(seq_ctx_list)
                 # Dense decoder layer - process concated hidden states
                 cat_hidden_states = decoder_layer(
                     cat_hidden_states,
                     position_embeddings=cat_position_embeddings,
                     seq_ctx=cat_seq_ctx,
                 )
             else:
-                if cat_hidden_states is not None and not moe_forawrd:
+                if not moe_forawrd:
```
[the remark this answers is not in the document]
Comment (HAOCHENYE) [reply xtuner/v1/model/moe/moe.py]: Fix this typo

File: xtuner/v1/model/moe/moe.py [hunk 3]
```diff
@@ -415,25 +404,24 @@ def _micro_batch_forward(
                     router_weights_list[i][f"layer{idx}"] = router_weights[i]
 
         # Apply final norm to all micro-batches
-        for i, hidden_states in enumerate(hidden_states_list):
-            hidden_states_list[i] = self.norm(hidden_states)
+        cat_hidden_states = torch.cat(hidden_states_list, dim=1)
```
[the remark this answers is not in the document]
Comment (HAOCHENYE) [reply xtuner/v1/model/moe/moe.py:407]: Agree

File: xtuner/v1/model/moe/moe.py [hunk 4]
```diff
@@ -415,25 +404,24 @@ def _micro_batch_forward(
...
-            if logits is not None:
-                logits_list.append(logits)
-            if extra_info:
-                moe_extra_info.append(extra_info)
+        cat_loss_ctx = CELossContext.cat(loss_ctx_list)
+        loss, (logits, extra_info) = self.lm_head(cat_hidden_states, cat_loss_ctx)  # type: ignore
 
         # Aggregate losses (mean across micro-batches)
-        output["loss"] = torch.stack(loss_list).sum() if loss_list else None
+        loss: torch.Tensor
```
[the remark this answers is not in the document]
Comment (HAOCHENYE) [reply xtuner/v1/model/moe/moe.py]: Please modify the type annotation of `LMHead.__call__`, the type of loss could be inferred as `toch.Tensor` automatically.

File: xtuner/v1/model/moe/moe.py [hunk 5]
```diff
@@ -415,25 +404,24 @@ def _micro_batch_forward(
...
+        output["loss"] = loss.sum()
+        moe_extra_info = ModelForwardExtraLogInfo()
+        if extra_info:
+            moe_extra_info.append(extra_info)
         output["extra_info"] = moe_extra_info
 
+        # Return logits for all micro-batches
+        final_logits = logits
```
Comment (HAOCHENYE) [inline xtuner/v1/model/moe/moe.py]: remove the variable `final_logits`

File: xtuner/v1/model/moe/moe.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (HIT-cwh) [reply xtuner/v1/model/moe/moe.py]: If loss_ctx is not None, the __call__ method will return a torch.Tensor; otherwise, it will return None. How should I modify the typehint? @HAOCHENYE

File: xtuner/v1/model/moe/moe.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (HIT-cwh) [reply xtuner/v1/model/moe/moe.py:407]: I don't think there will be a moe model where all its layers are dense layers. haha

File: xtuner/v1/model/moe/moe.py (hunk 5 above)
Comment (HIT-cwh) [reply xtuner/v1/model/moe/moe.py]: When constructing MoEModelOutputs, final_logits will be utilized.
